packages/std/src/memory.rs

```diff
@@ -116,55 +54,167 @@ unsafe impl RegionSource for Vec<u8> {
     }
 }
 
-unsafe impl<T: ?Sized> RegionSource for &T
+mod sealed {
+    pub trait Sealed: 'static {}
+
+    impl Sealed for super::Owned {}
+
+    impl Sealed for super::Borrowed {}
+}
+
+pub trait Ownership: sealed::Sealed + 'static {}
+
+impl<T> Ownership for T where T: sealed::Sealed {}
+
+pub struct Owned {}
+
+pub struct Borrowed {}
```

Comment:
```suggestion
pub struct Owned;

pub struct Borrowed;
```

Can we do this instead?